File: MonoScripting/Api/EngineClass/EntityComponents/MaterialComponent.cs
using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

public class MaterialComponent : IEntityComponent {

    public MaterialComponent(ulong handle) : base(handle)
    {
    }

    public void SetMaterial(string resource_name){
        i_SetMaterial(handle, resource_name);
    }

    void SetMaterial(Resource resource){
        i_SetMaterial(handle, resource.GetHandle());
    }

    bool IsCastShadows(){
        return i_IsCastShadows(handle);
    }

    void SetCastShadows(bool cast){
        i_SetCastShadows(handle, cast);
    }

    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern string i_SetMaterial(ulong handle, string material_name);

    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern string i_SetMaterial(ulong handle, ulong resource_handle);


    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern bool i_IsCastShadows(ulong handle);
    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern void i_SetCastShadows(ulong handle, bool cast);
}

File: Engine/MonoScripting/Api/EntityComponents/MaterialComponent.cpp
#include "../ApiBindings.hpp"
#include <Scene/EntityComponents/MaterialComponent.hpp>
#include <MonoScripting/MonoScriptingLayer.hpp>

using namespace VSGE;

static void SetMaterial(MaterialComponent* comp, Resource* resource) {
	comp->SetMaterialId(resource->getId());
}

static Material* GetMaterial(MaterialComponent* comp) {
	return comp->GetMaterial();
}

static bool IsCastShadows(MaterialComponent* comp) {
	return comp->IsCastShadows();
}

static void SetCastShadows(MaterialComponent* comp, bool cast) {
	return comp->SetCastShadows(cast);
}

void VSGE::BindMaterial() {
	MonoScriptingLayer::AddInternalCall("MaterialComponent::i_SetMaterial(ulong,ulong)", SetMaterial);
	MonoScriptingLayer::AddInternalCall("MaterialComponent::i_GetMaterialHandle(ulong)", GetMaterial);

	MonoScriptingLayer::AddInternalCall("MaterialComponent::i_IsCastShadows(ulong)", IsCastShadows);
	MonoScriptingLayer::AddInternalCall("MaterialComponent::i_SetCastShadows(ulong,bool)", SetCastShadows);
}